job/jobcache.go

```diff
@@ -1,36 +1,86 @@
 package job
 
 import (
+	"os"
+	"os/signal"
 	"sync"
+	"time"
 )
 
-var (
-	AllJobs = &JobCache{
-		Jobs:   make(map[string]*Job),
-		rwLock: sync.Mutex{},
+type JobCache interface {
+	Get(id string) *Job
+	GetAll() map[string]*Job
+	Set(j *Job)
+	Delete(id string)
+	Persist() error
+}
+
+type MemoryJobCache struct {
+	Jobs            map[string]*Job
+	rwLock          sync.Mutex
+	jobDB           JobDB
+	persistWaitTime time.Duration
+}
+
+func NewMemoryJobCache(jobDB JobDB, persistWaitTime time.Duration) *MemoryJobCache {
+	if persistWaitTime == time.Duration(0) {
+		persistWaitTime = 5 * time.Second
 	}
-)
+	return &MemoryJobCache{
+		Jobs:            map[string]*Job{},
+		rwLock:          sync.Mutex{},
+		jobDB:           jobDB,
+		persistWaitTime: persistWaitTime,
+	}
+}
+
+func (c *MemoryJobCache) Init() {
+	// Prep cache
+	allJobs, err := c.jobDB.GetAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, v := range allJobs {
+		v.StartWaiting(c)
+		c.Set(v)
+	}
+
+	// Occasionally, save items in cache to db.
+	go c.PersistEvery(c.persistWaitTime)
+
+	// Process-level defer for shutting down the db.
+	ch := make(chan os.Signal, 1)
+	signal.Notify(ch, os.Interrupt, os.Kill)
+	go func() {
+		s := <-ch
+		log.Info("Process got signal: %s", s)
+		log.Info("Shutting down....")
 
-type JobCache struct {
-	Jobs   map[string]*Job
-	rwLock sync.Mutex
+		// Persist all jobs to database
+		c.Persist()
+
+		// Close the database
+		c.jobDB.Close()
+
+		os.Exit(1)
+	}()
 }
 
-func (c *JobCache) Get(id string) *Job {
+func (c *MemoryJobCache) Get(id string) *Job {
 	c.rwLock.Lock()
 	defer c.rwLock.Unlock()
 
 	return c.Jobs[id]
 }
 
-func (c *JobCache) GetAll() map[string]*Job {
+func (c *MemoryJobCache) GetAll() map[string]*Job {
 	c.rwLock.Lock()
 	defer c.rwLock.Unlock()
 
 	return c.Jobs
 }
```

Comment:
@ajvb Here the lock isn't doing a whole lot for you.  This is only safe to do if the caller isn't going to read from or write to `c.Jobs` :-), which is basically never.  What I'm saying is, the caller's access to the returned data structure aren't lock-protected, so this is still racy.